### ISSUE TITLE
Children of children and recursive queries

### ISSUE THREAD

**Issue by guaraqe** [guaraqe/csdc#82]
The graph of units in the DAO is supposed to be a DAG, and we want to be able to have operations that work not only on units themselves, but also on their whole subtree. There are at least three operations that we need to support before reaching these features:

- Detecting for cycles, which are forbidden. This is not an artificial constraint, but necessary if we want to have bottom-up elections. This operation will be done each time a unit-unit link is created.
- Detection of whether a person is a direct or indirect member of an unit. This allows for some degree of openness and privacy in units. This will be done each time a person visits a unit.
- Collection of all direct and indirect members of a unit, so that notifications can be sent.

This is something I never did before in PostgreSQL, so I started investigating this in #81.

While searching for a solution, I found a post from [Mergify](https://blog.mergify.com/cycle-detection-in-postgresql/), which gives a solution for cycles, and it mentions [this blog post](https://www.depesz.com/2021/02/04/waiting-for-postgresql-14-search-and-cycle-clauses/?ref=blog.mergify.com). It seems that in PostgreSQL 14 there are clauses to help with recursive queries an, namely `SEARCH` and `CYCLE`. 

Here I will try to organize my thoughts on how to use these to support these operations. The base recursive query I will use it the one from #81:

```sql
WITH RECURSIVE descendants(id, parent, child, level) AS (
    SELECT id, parent, child, 1 as level
    FROM subparts
    UNION
    SELECT d.id, d.parent, s.child, d.level + 1
    FROM descendants d JOIN subparts s ON d.child=s.parent
)
```

### Cycle detection

The [documentation](https://www.postgresql.org/docs/current/queries-with.html) shows how PostgreSQL can detect cycles itself, by adding a column `is_cycle` to the request:

```sql
WITH RECURSIVE descendants(id, parent, child, level) AS (
    SELECT id, parent, child, 1 as level
    FROM subparts
    UNION
    SELECT d.id, d.parent, s.child, d.level + 1
    FROM descendants d JOIN subparts s ON d.child=s.parent
) CYCLE child SET is_cycle USING path
```

which will also return the `path` in case we need to show what is the path. The solution would be to add a `CONSTRAINT` to the database that check if a result `EXISTS` when there is a `is_cycle = TRUE`.

### Indirect membership

Direct membership is very easy to check due to the `members` table. However, indirect membership is more complex. We can reduce this problem to the present scope as: is one of the units of which the person is member an indirect child of the target unit?

A simplistic solution of this would be to execute the request above, and join with the person's units:

```sql
SELECT *
FROM descendants JOIN members ON descendants.child = members.unit
WHERE descendants.parent = $target_unit AND members.person = $target_person
```

and check whether the result exists or not. 

### Indirect member collection

The approach in #81 is sufficient for that. However, if we want to have better control about the order of traversal, we can use, for example, for depth first searches:

```sql
WITH RECURSIVE descendants(id, parent, child, level) AS (
    SELECT id, parent, child, 1 as level
    FROM subparts
    UNION
    SELECT d.id, d.parent, s.child, d.level + 1
    FROM descendants d JOIN subparts s ON d.child=s.parent
) SEARCH DEPTH FIRST BY path SET order
```

which will also add the `path` and `order` columns to the request.

**Comment by guaraqe** [comment]
Cycle detection is done in https://github.com/guaraqe/csdc/pull/81.

**Comment by guaraqe** [comment]
I implemented each one of these in #81. Now we know how to detect indirect membership, and how to retrieve indirect members, so we can start making use of it.
